### PR TITLE
SetDigest jaccard_index implementation Issue

### PR DESCRIPTION
**Pull request by Akanksha-kedia**
SetDigest jaccard_index implementation Issue

https://en.wikipedia.org/wiki/Jaccard_index

 

For below query the jaccard_index should be equal to 0.3333333333333333 but the output is coming as 0.5 which is incorrect 

SELECT jaccard_index(make_set_digest(value), make_set_digest(value1)) FROM (VALUES ('abc', 'def'),('ee', 'abc')) T(value,value1);

### CODE REVIEW

**Comment by Akanksha-kedia** [comment]
@tdcmeehan  please review

**Comment by Akanksha-kedia** [comment]
@tdcmeehan please review and merge

**Comment by tdcmeehan** [comment]
Could you please add some unit tests?

**Comment by Akanksha-kedia** [comment]
@tdcmeehan HAVE ADDED TEST CASES

**Comment by Akanksha-kedia** [comment]
The Jaccard Index is a measure used for comparing the similarity and diversity of sample sets. It is calculated as the intersection divided by the union of the sample sets.

However, in this case, you have provided pairs of values, not sets. The Jaccard Index is not typically used for pairs of values like this.

If we consider your pairs as sets, then:

For (1,1) and (NULL,2), there is no common element, so the Jaccard Index is 0.
For (1,1) and (2,3), there is no common element, so the Jaccard Index is 0.
For (1,1) and (NULL,4), there is no common element, so the Jaccard Index is 0.
For (NULL,2) and (2,3), there is one common element (2), so the Jaccard Index is 1/3 = 0.33.
For (NULL,2) and (NULL,4), there is one common element (NULL), so the Jaccard Index is 1/3 = 0.33.
For (2,3) and (NULL,4), there is no common element, so the Jaccard Index is 0.


If we treat these as two sets:

Set A = {1, null, 2, null} = {1, null, 2} (since sets only contain unique elements) Set B = {1, 2, 3, 4}

The intersection of A and B (elements common to both sets) is {1, 2} which has 2 elements.

The union of A and B (all unique elements in both sets) is {1, null, 2, 3, 4} which has 5 elements.

Therefore, the Jaccard Index, which is the size of the intersection divided by the size of the union, would be 2/5 = 0.4.

**Comment by Akanksha-kedia** [comment]
@tdcmeehan please review

**Comment by tdcmeehan** [comment]
I do not have enough knowledge about jaccard index to be confident in merging it.  Let's get some additional eyes--I'll find other reviewers.

File: presto-main/src/main/java/com/facebook/presto/type/setdigest/SetDigest.java [hunk 1]
```diff
@@ -165,22 +165,16 @@ public static long exactIntersectionCardinality(SetDigest a, SetDigest b)
...
         LongSortedSet minUnion = new LongRBTreeSet(a.minhash.keySet());
         minUnion.addAll(b.minhash.keySet());
 
         int intersection = 0;
-        int i = 0;
         for (long key : minUnion) {
             if (a.minhash.containsKey(key) && b.minhash.containsKey(key)) {
                 intersection++;
             }
-            i++;
```
Comment (jaystarshot) [inline presto-main/src/main/java/com/facebook/presto/type/setdigest/SetDigest.java:178]: why is this removed? this was a good optimization

File: presto-main/src/main/java/com/facebook/presto/type/setdigest/SetDigest.java [hunk 2]
```diff
@@ -165,22 +165,16 @@ public static long exactIntersectionCardinality(SetDigest a, SetDigest b)
...
-            if (i >= sizeOfSmallerSet) {
-                break;
-            }
         }
-        return intersection / (double) sizeOfSmallerSet;
+        return intersection / (double) minUnion.size();
```
Comment (jaystarshot) [inline presto-main/src/main/java/com/facebook/presto/type/setdigest/SetDigest.java:177]: I think the fix is only needed for this line

File: presto-main/src/main/java/com/facebook/presto/type/setdigest/SetDigest.java (hunk 1 above)
Comment (jaystarshot) [inline presto-main/src/main/java/com/facebook/presto/type/setdigest/SetDigest.java:178]: I was mistaken, small optimization - maybe we can break if (intersection >= sizeofSmallerSet) ?

**Comment by jaystarshot** [comment]
Lgtm

File: presto-main/src/main/java/com/facebook/presto/type/setdigest/SetDigest.java [hunk 3]
```diff
@@ -165,22 +165,16 @@ public static long exactIntersectionCardinality(SetDigest a, SetDigest b)
 
     public static double jaccardIndex(SetDigest a, SetDigest b)
     {
-        int sizeOfSmallerSet = Math.min(a.minhash.size(), b.minhash.size());
         LongSortedSet minUnion = new LongRBTreeSet(a.minhash.keySet());
         minUnion.addAll(b.minhash.keySet());
 
         int intersection = 0;
-        int i = 0;
         for (long key : minUnion) {
```
Comment (ClarenceThreepwood) [inline presto-main/src/main/java/com/facebook/presto/type/setdigest/SetDigest.java:172]: This is iterating over all elements in minUnion and then checking for the element's membership in each input set.
How about just iterating over the elements in the smaller set among the two, and then checking if the element exists  in the other?

File: presto-main/src/main/java/com/facebook/presto/type/setdigest/SetDigest.java (hunk 2 above)
Comment (ClarenceThreepwood) [reply presto-main/src/main/java/com/facebook/presto/type/setdigest/SetDigest.java:177]: +1

**Review by ClarenceThreepwood — CHANGES_REQUESTED** [review]

**Comment by ClarenceThreepwood** [comment]
Also please fix checkstyle and failing tests

**Comment by hmadison** [comment]
[Following up from Slack](https://prestodb.slack.com/archives/CLYMEEJ6S/p1695740671700649?thread_ts=1695737791.924169&cid=CLYMEEJ6S), while the implementation looks to be correct I'm concerned that this is a breaking change. Would making a three arity function for jaccard_index make more sense and leave the two arity function with the current implementation until there is some kind of deprecation cycle?

**Comment by Akanksha-kedia** [comment]
Yes, it makes sense.But lets take more understanding from the community
people.

**Comment by Akanksha-kedia** [comment]
@ClarenceThreepwood please review

File: presto-main/src/main/java/com/facebook/presto/type/setdigest/SetDigest.java [hunk 4]
```diff
@@ -183,6 +183,25 @@ public static double jaccardIndex(SetDigest a, SetDigest b)
         return intersection / (double) sizeOfSmallerSet;
     }
 
+    public static double jaccardIndex(SetDigest a, SetDigest b)
+    {
+        int sizeOfSmallerSet = Math.min(a.minhash.size(), b.minhash.size());
```
Comment (ClarenceThreepwood) [inline presto-main/src/main/java/com/facebook/presto/type/setdigest/SetDigest.java:188]: What do you think of something like the following instead

    public static double jaccardIndex(SetDigest a, SetDigest b)
    {
        SetDigest smaller;
        SetDigest larger;
        if (a.minhash.size() < b.minhash.size())
        {
            smaller = a;
            larger = b;
        }
        else {
            smaller = b;
            larger = a;
        }

        LongSortedSet smallerSet = new LongRBTreeSet(smaller.minhash.keySet());
        LongSortedSet largerSet = new LongRBTreeSet(larger.minhash.keySet());

        int intersection = 0;
        for (long key : smallerSet) {
            if (largerSet.contains(key)) {
                intersection++;
            }
        }
        LongSortedSet union = new LongRBTreeSet(largerSet);
        union.addAll(smallerSet);

        return intersection / (double) union.size();
    }

**Comment by Akanksha-kedia** [comment]
-------------------------------------------------------
[INFO]  T E S T S
[INFO] -------------------------------------------------------
[INFO] Running com.facebook.presto.type.setdigest.TestSetDigest
[INFO] Tests run: 4, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 9.796 s - in com.facebook.presto.type.setdigest.TestSetDigest
[INFO] 
[INFO] Results:
[INFO] 
[INFO] Tests run: 4, Failures: 0, Errors: 0, Skipped: 0
[INFO] 
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------


Tests run: 4, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 14.264 s - in com.facebook.presto.sql.query.TestSetDigestFunctions
[INFO] 
[INFO] Results:
[INFO] 
[INFO] Tests run: 4, Failures: 0, Errors: 0, Skipped: 0
[INFO] 
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
 @tdcmeehan @ClarenceThreepwood

**Comment by tdcmeehan** [comment]
@Akanksha-kedia please rebase to pick up test fixes.

**Comment by Akanksha-kedia** [comment]
> test

i did not understand which tests i need to pick

**Comment by tdcmeehan** [comment]
Please just register one function, and switch it with a config property.  Here's an example: https://github.com/prestodb/presto/blob/7a6f3768b0f5e2ca9c6052cd1e73f28420874aaa/presto-main/src/main/java/com/facebook/presto/metadata/BuiltInTypeAndFunctionNamespaceManager.java#L958

**Comment by tdcmeehan** [comment]
Please reopen when you've updated the PR to address the feedback, thanks!
